ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java

```diff
@@ -0,0 +1,211 @@
+    for (Path obsolete : dir.getObsolete()) {
+      FileStatus stat = dir.getFs().getFileStatus(obsolete);
+      if (System.currentTimeMillis() - stat.getModificationTime() >= checkThresholdInSec * 1000) {
+        numObsoleteDeltas++;
+      }
+    }
+    return numObsoleteDeltas;
+  }
+
+  public static void close() {
+    if (getInstance() != null) {
+      getInstance().executorService.shutdownNow();
+    }
+  }
+
+  private void configure(HiveConf conf){
+    deltasThreshold = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_DELTA_NUM_THRESHOLD);
+    obsoleteDeltasThreshold = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_OBSOLETE_DELTA_NUM_THRESHOLD);
+    checkThresholdInSec = HiveConf.getTimeVar(conf,
+        HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_DELTA_CHECK_THRESHOLD, TimeUnit.SECONDS);
+
+    initMetricsCache(conf);
+    long reportingInterval = HiveConf.getTimeVar(conf,
+        HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_REPORTING_INTERVAL, TimeUnit.SECONDS);
+
+    ThreadFactory threadFactory =
+      new ThreadFactoryBuilder()
+        .setDaemon(true)
+        .setNameFormat("DeltaFilesMetricReporter %d")
+        .build();
+    executorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    executorService.scheduleAtFixedRate(
+        new ReportingTask(), 0, reportingInterval, TimeUnit.SECONDS);
+    LOG.info("Started DeltaFilesMetricReporter thread");
+  }
+
+  private void initMetricsCache(HiveConf conf) {
+    maxCacheSize = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_MAX_CACHE_SIZE);
+    long duration = HiveConf.getTimeVar(conf,
+        HiveConf.ConfVars.HIVE_TXN_ACID_METRICS_CACHE_DURATION, TimeUnit.SECONDS);
+
+    Comparator<Pair<String, Integer>> c = Comparator.comparing(Pair::getValue);
+    deltaTopN = new PriorityBlockingQueue<>(maxCacheSize, c);
+    obsoleteDeltaTopN = new PriorityBlockingQueue<>(maxCacheSize, c);
+
+    deltaCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(duration, TimeUnit.SECONDS)
+      .maximumSize(maxCacheSize)
+      .removalListener(notification -> removalPredicate(deltaTopN, notification))
+      .softValues()
+      .build();
+
+    obsoleteDeltaCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(duration, TimeUnit.SECONDS)
+      .maximumSize(maxCacheSize)
+      .removalListener(notification -> removalPredicate(obsoleteDeltaTopN, notification))
+      .softValues()
+      .build();
+  }
+
+  private boolean removalPredicate(BlockingQueue<Pair<String, Integer>> deltaTopN, RemovalNotification notification) {
+    return deltaTopN.removeIf(item -> item.getKey().equals(notification.getKey()));
+  }
+
+  private final class ReportingTask implements Runnable {
+    @Override
+    public void run() {
+      Metrics metrics = MetricsFactory.getInstance();
+      if (metrics != null) {
+          obsoleteDeltaCache.cleanUp();
+          metrics.addGauge("compaction_num_obsoleteDeltas",
```

Comment:
nit: CamelCase mixed with snake_case

Also the metric names should be constants.